packages/workflow/src/workflow.ts

```diff
@@ -1356,21 +1372,98 @@ export function upsertSearchAttributes(searchAttributes: SearchAttributes): void
     throw new Error('searchAttributes must be a non-null SearchAttributes');
   }
 
-  activator.pushCommand({
-    upsertWorkflowSearchAttributes: {
-      searchAttributes: mapToPayloads(searchAttributePayloadConverter, searchAttributes),
-    },
-  });
+  if (Array.isArray(searchAttributes)) {
+    // Typed search attributes
+    activator.pushCommand({
+      upsertWorkflowSearchAttributes: {
+        searchAttributes: typedMapToPayloads<string, TypedSearchAttributeUpdateValue>(
+          typedSearchAttributePayloadConverter,
+          Object.fromEntries(searchAttributes.map(([k, v]) => [k.name, v]))
+        ),
+      },
+    });
 
-  activator.mutateWorkflowInfo((info: WorkflowInfo): WorkflowInfo => {
-    return {
-      ...info,
-      searchAttributes: {
-        ...info.searchAttributes,
-        ...searchAttributes,
+    activator.mutateWorkflowInfo((info: WorkflowInfo): WorkflowInfo => {
+      const newSearchAttributes: SearchAttributes = {};
+      for (const pair of searchAttributes) {
+        const [k, v] = pair;
+        if (v === null) {
+          // If the value is null (i.e. delete), set the value to an empty list.
+          newSearchAttributes[k.name] = [];
+        } else {
+          const typedValue = v[1];
+          newSearchAttributes[k.name] = Array.isArray(typedValue) ? typedValue : ([typedValue] as SearchAttributeValue);
+        }
+      }
+      return {
+        ...info,
+        searchAttributes: {
+          ...info.searchAttributes,
+          ...newSearchAttributes,
+        },
+        typedSearchAttributes: info.typedSearchAttributes.updateAttributes(searchAttributes),
+      };
+    });
+  } else {
+    // Legacy search attributes
+    activator.pushCommand({
+      upsertWorkflowSearchAttributes: {
+        searchAttributes: mapToPayloads(searchAttributePayloadConverter, searchAttributes),
       },
-    };
-  });
+    });
+
+    activator.mutateWorkflowInfo((info: WorkflowInfo): WorkflowInfo => {
+      // Use updateSearchAttributes to get a deep copy of the current state.
+      let typedSearchAttributes = info.typedSearchAttributes.updateAttributes([]);
+      for (const [k, v] of Object.entries(searchAttributes)) {
+        const typedKey = TypedSearchAttributes.getKeyFromUntyped(k, v);
+
+        // Unable to discern the typing of the key, skip.
+        if (!typedKey) {
+          continue;
+        }
+
+        // Delete the key if the value is undefined/null/empty list.
+        if (!v || (Array.isArray(v) && v.length === 0)) {
+          const pair = [typedKey, null];
+          if (isTypedSearchAttributeUpdatePair(pair)) {
+            typedSearchAttributes = typedSearchAttributes.updateAttributes([pair]);
+          }
+          continue;
+        }
+
+        let newValue: unknown = v;
+        if (typedKey.type !== SearchAttributeType.KEYWORD_LIST) {
+          // Skip if we found multiple values for a non-list key.
+          if (Array.isArray(v) && v.length > 1) {
```

Comment:
Remove from `typedSearchAttributes` in that case. That's better than having incohenrencies.

Reply:
Ditto the previous comment on removing from `typedSearchAttributes`.

In any case, I've removed this check and a couple other validation checks in the `upsert` logic. If we can determine a valid type for the given input then these additional checks are redundant.


I've also left a short comment in the `upsert` logic that covers the basic cases, thought it might be useful though the logic is not complicated. Can remove if you think unnecessary. Testing these cases (i.e. multi-value input for non keyword list, invalid objects/primitives) is difficult given that server will error if you try to do this.

Reply:
Removed the comment. Code is pretty readable and doesn't really warrant it.